### PR TITLE
Suggestions and Activity Pub spec

### PR DESCRIPTION
**Pull request by colin-axner**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,7 +10,7 @@ A **self-regulating** community is a community in which the users themselves col
...
 
-A possible reaction to the current centralized Internet platforms might be to build a completely decentralized, censorship-resistant, anonymous social media platform. Such a platform certainly doesn't suffer from a central authority abusing its power, however it is not self-regulating since there is no way to enforce any rules on such a system. In fact it barely fits the definition of a community above. Such a system can easily devolve into chaotic, uncivil pandemonium of the sort we see in less-scrupulously moderated social media platforms already existing today.
```
Comment (colin-axner) [inline README.md:13]: I thought the wording became slightly strong, I like keeping it neutral which was well done with the facebook example

File: core/activity_pub.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+
+## Public/Private Key Authentication
+
+ActivityPub does not specify a specific authentication scheme, however most ActivityPub communities use OAuth2.0 or something similar which involves a traditional username/password scheme. Since all of the community state must be public and shareable between hosts, we cannot rely on username/password schemes or any scheme that gives hosts privileged information on a user’s authentication info. Thus, users must authenticate by proving ownership of a secret private key that corresponds to the public key that is publicly registered on a blockchain.
+
+## URI and DID
```
Comment (colin-axner) [inline core/activity_pub.md]: This needs a lot more, but I need to dive deeper into dids first

File: dogma/goals.md [hunk 3]
```diff
@@ -13,12 +13,13 @@
 7. In order to efficiently and effectively enforce community rules, members of the community may appoint trusted users to specific roles that help steward the community (eg admitting new members, enforcing rules, etc). However, they may also replace these appointees at any time by voting them out and replacing with a new candidate. 
 8. Any malicious authority must not be able to use their power to suppress the ability of the community to overthrow them. For example, a moderator should not be able to ban all users that vote against them before they can vote them out.
 9. *A user who wishes to individually exit the community should be able to transfer their established identity to a different community.*
-10. *Community governance should be a liquid democracy where users can delegate their voting power to other users who can vote on their behalf*
+10. *Users in different communities should be capable of communicating with one another (given both parties consent consent).*
```
Comment (colin-axner) [inline dogma/goals.md]: interoperability is a good goal

File: README.md (hunk 1 above)
Comment (AdityaSripal) [reply README.md:13]: I agree the wording in this passage is strong, but I do feel that such a platform would not actually be a community.

Maybe we should reword "easily devolve into chaotic, uncivil pandemonium" instead to tone it down while still conveying the same point.

File: core/activity_pub.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# Activity Pub Compliance
+
+This document specifies the extensions to Fora required for a community to be compliant with the [Activity Pub](https://www.w3.org/TR/activitypub/) specification. Without this extension, the Fora kernel will not be capable of interacting with other Activity Pub implementations.
+
+## Introduction
+
+"it's just activity pub with governance"
+
+If we examine the Fora kernel through the lens of Activity Pub we find numerous similarities. Each user is a Client and each host is a Server, as defined by the Activity Pub specification. In fact, the only differences are a layer of governance settled on top of Activity Pub, along with additional properties between clients and servers, and servers and servers. The primary property added is the right to collective identification as a single community. Different servers with different sets of users can now collectively agree to be identified as the same community thereby sharing their state and community membership.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
If we examine the Fora kernel through the lens of Activity Pub we find numerous similarities. Each user is a Client and each host is a Server, as defined by the Activity Pub specification. In fact, the only differences are a layer of governance settled on top of Activity Pub, along with additional properties between clients and servers, and servers and servers. The primary property added is the right to collective identification as a single community. Different servers with different sets of clients can now collectively agree to be identified as the same community thereby sharing their state and community membership.
```

Lets keep uniform terminology to avoid confusion. When we're talking in ActivityPub context, let's always use the same "Server/Client" pair terminology.

Different servers with different sets of clients may still all be part of the same set of users in a fora community.

File: core/activity_pub.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+The usefulness of this conceptual nesting is addressing the **right to individual exit** and the **right to collective exit**. 
+
+The former can be relatively easily addressed in Activity Pub by using flexible URI's such as DID's. This would give any user power to exit a community without losing it's account information (content may be lost, but followers/following lists would not).
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
The former can be relatively easily addressed in Activity Pub by using flexible URI's such as DID's. This would give any user power to exit a community without losing its account information (content may have to be manually ported, but followers/following lists would automatically follow the user).
```

File: core/activity_pub.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+
+## Public/Private Key Authentication
+
+ActivityPub does not specify a specific authentication scheme, however most ActivityPub communities use OAuth2.0 or something similar which involves a traditional username/password scheme. Since all of the community state must be public and shareable between hosts, we cannot rely on username/password schemes or any scheme that gives hosts privileged information on a user’s authentication info. Thus, users must authenticate by proving ownership of a secret private key that corresponds to the public key that is publicly registered on a blockchain.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
ActivityPub does not specify a particular authentication scheme, however most ActivityPub communities use OAuth2.0 or something similar which involves a traditional username/password scheme. Since all of the community state must be public and shareable between hosts, we cannot rely on username/password schemes or any scheme that gives hosts privileged information on a user’s authentication info. Thus, users must authenticate by proving ownership of a secret private key that corresponds to the public key that is publicly registered on a blockchain.
```

File: core/activity_pub.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+## URI and DID
+
+URIs (Uniform Resource Identifier) are used within Activity Pub to identify different servers. 
+
+Fora will use [decentralized identifiers](https://www.w3.org/TR/did-core/), to identify both users and platforms. These DIDs must be registered on a blockchain that the platform can query identities from. Users will have a DID so they can self-attest to their identity on the blockchain and then use that DID in any Fora community they wish. They can also retain this identity even as they move to different communities. The DID document for each user must contain the current public key they will use sign messages sent from their Fora user account. 
+
+Fora implementations which rely on the public key registered with a DID instead of registering a public key with a governance module, will benefit from dynamic swapping of public keys. 
+
+Meanwhile communities will also have DIDs that identify them to their users. This DID will also be registered on the blockchain as a government. The DID will resolve to the current hosts of the platform. Any governance decisions which affects the set of hosts will update the DID resolving result to output the new set of hosts.
```
Comment (AdityaSripal) [inline core/activity_pub.md]: It should be clear that a community does not have a defined set of hosts. Anyone can be a host.

What we want to say here is that a community may define a default set of hosts as a UX convenience.

```suggestion
Meanwhile communities will also have DIDs that identify them to their users. This DID will also be registered on the blockchain as a government. The DID may include a default set of "seed" hosts to allow users to conveniently find and connect to the community. Any governance decisions which affects the set of hosts will update the DID resolving result to output a new set of hosts.

Note: There may be many valid, honest hosts that exist outside of this community-curated set.
```

File: core/activity_pub.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+## Server
+
+In addition to maintaining the full state of the community, servers will listen to events on the ActivityPub network to record new interactions. Servers must be connected to the appropriate blockchains or DID resolution services in order to track and verify the DID of users in the community.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
In addition to maintaining the full state of the community, servers will listen to events on the ActivityPub network to record and propagate new interactions. Servers must be connected to the appropriate blockchains or DID resolution services in order to track and verify the DID of users in the community.
```

File: core/activity_pub.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+## Limitations
+
+Since any content posted to a Fora community must contend with the possibility of byzantine host behaviour, the Fora protocol cannot make any guarantees regarding private posts that are only readable to a subset of users. Thus, Fora will only support Public posts and direct messages. Public posts will be sent in plaintext and signed by the sender, while direct messages to users can be end-to-end encrypted given that all users have a Public/Private key as mentioned above (though a different pair would be used for DMs). This ensures that hosts have no ability to attack the confidentiality or integrity of user-sent messages. They can only refuse to relay the messages, though this is grounds for enforcers to remove a host.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
Since any content posted to a Fora community must contend with the possibility of byzantine host behaviour, the Fora protocol cannot make any guarantees regarding private posts that are only readable to a subset of users. Thus, Fora will only support Public posts and direct messages. Public posts will be sent in plaintext and signed by the sender, while direct messages to users can be end-to-end encrypted given that all users have a Public/Private key as mentioned above (though a different pair would be used for DMs). This ensures that hosts have no ability to attack the confidentiality or integrity of user-sent messages. They can only refuse to relay the messages, though the message may simply get relayed through a different path.
```

There isn't any absolute way to "remove" a host, bad hosts must simply be circumvented.

File: core/activity_pub.md (hunk 6 above)
Comment (AdityaSripal) [reply core/activity_pub.md]: "specify a specific" sounds awkward

File: dogma/goals.md (hunk 3 above)
Comment (AdityaSripal) [inline dogma/goals.md]:
```suggestion
10. *Users in different communities should be capable of communicating with one another (given both parties consent).*
```

File: core/core.md [hunk 10]
```diff
@@ -8,7 +8,7 @@ In order to understand the Fora protocol, it is useful to define the entities in
...
+**Host**: As mentioned above, a host is a member that maintains the full state of the community. They will receive messages from other users of the community over a gossip network and update their own state, as well as gossip any updates to the rest of the network. They may or may not serve as an access point for Light users. Hosts will listen to events on the blockchain in order to keep track of any decisions made there. An honest Host will delete any banned content from their local copy of the community and ignore any banned user by refusing to accept and propogate their messages.
```
Comment (AdityaSripal) [inline core/core.md]: I get that we may want to separate out the ActivityPub specifics from the core document. But Hosts must listen to blockchain decisions as well as community interactions. Maybe rephrase to:

```suggestion
**Host**: As mentioned above, a host is a member that maintains the full state of the community. They will receive messages from other users of the community over a gossip network and update their own state, as well as gossip any updates to the rest of the network. They may or may not serve as an access point for Light users. Hosts will record and propogate community interactions, and they must also listen to events on the blockchain in order to abide by any decisions made there. An honest Host will delete any banned content from their local copy of the community and ignore any banned user by refusing to accept and propogate their messages.
```

**Review by AdityaSripal — CHANGES_REQUESTED** [review]
Great start! I like the structure of the ActivityPub doc, have some wording and conceptual requests.

File: README.md (hunk 1 above)
Comment (colin-axner) [reply README.md:13]: I think that part is fine, my main concern was `In fact it barely` in conjunction with the above. The `In fact it barely` feels like 2 strong terms `in fact` and `it barely`. I'm fine with my currently proposed suggestion of `it does not necessarily` but open to other suggestions if the original meaning is lost

File: core/activity_pub.md [hunk 11]
```diff
@@ -6,35 +6,46 @@ This document specifies the extensions to Fora required for a community to be co
...
 
 
-## Public/Private Key Authentication
+## URI and DID
 
-ActivityPub does not specify a specific authentication scheme, however most ActivityPub communities use OAuth2.0 or something similar which involves a traditional username/password scheme. Since all of the community state must be public and shareable between hosts, we cannot rely on username/password schemes or any scheme that gives hosts privileged information on a user’s authentication info. Thus, users must authenticate by proving ownership of a secret private key that corresponds to the public key that is publicly registered on a blockchain.
+URIs (Uniform Resource Identifier) are used within Activity Pub to identify different servers and clients.  
 
-## URI and DID
+Fora will use [decentralized identifiers](https://www.w3.org/TR/did-core/), to identify both clients and communities.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
Fora will use [decentralized identifiers](https://www.w3.org/TR/did-core/), to identify both members and communities. 
```

Again let's stick to Server/Client for ActivityPub and Community/Member. See #2

File: core/activity_pub.md [hunk 12]
```diff
@@ -6,35 +6,46 @@ This document specifies the extensions to Fora required for a community to be co
...
+The DIDs of a client must be recorded on a verifiable data registry, typically this will be a public blockchain or ledger. All communities will rely on client DIDs to resolve the authentication method used by a particular client. Communities may choose not to rely on DIDs for authentication of governance module interaction (NOTE: the authentication method of a client may be different from its corresponding controller, the latter would be used to authenticate a governance vote). In this case, public keys for each client controller must be registered on the community. If the communities chooses to use on-chain DID verification, the client DID controllers must be verifiable via a communication protocol such as InterBlockchain Communication.
```
Comment (AdityaSripal) [inline core/activity_pub.md]: "(NOTE: the authentication method of a client may be different from its corresponding controller, the latter would be used to authenticate a governance vote)"

Could you clarify what you mean here? Do you mean that the PubKey used to send ActivityPub messages could be different than the pubkey used to send governance votes

File: core/activity_pub.md [hunk 13]
```diff
@@ -6,35 +6,46 @@ This document specifies the extensions to Fora required for a community to be co
...
+Communities will also have DIDs that identify them to their users. This DID will also be recorded on a blockchain as a government. The DID Document for this DID will be deterministically generated based on the community parameters that exist at the block height the DID resolving function is being executed.  All governance decisions will be recorded on the DID Document. The DID controller will be the community members and the verification method to update the DID Document will be the passing of a governance proposal. The DID may include a default set of "seed" hosts to allow users to conveniently find and connect to the community which will be set under the service endpoint section of a DID Document. Additional parameters will be added to the DID Document which indicate a list of admittors, enforcers, members, governance decisions and any other relevant information.
```
Comment (AdityaSripal) [inline core/activity_pub.md]: Does the service endpoint only return the list of hosts? Can I use the DID to query for the admittors, enforcers, community parameters, constitution, etc?

**Review by AdityaSripal — CHANGES_REQUESTED** [review]
Nice work! Just have a few clarifications and minor suggestions, then this is good to merge

File: README.md (hunk 1 above)
Comment (AdityaSripal) [reply README.md:13]: Sure, my only thinking here is that I want to convey that this system is not a community unambiguosly.

I'm good with: `In fact it does not`

File: core/activity_pub.md (hunk 12 above)
Comment (colin-axner) [reply core/activity_pub.md]: Correct. There exists [authorization](https://www.w3.org/TR/did-core/#did-controller) in a DID Document
```
Authorization is the mechanism used to state how operations are performed on behalf of the DID subject. A DID controller is authorized to make changes to the respective DID document. 
```

As well as [authentication](https://www.w3.org/TR/did-core/#authentication). 
```
The authentication verification relationship is used to specify how the DID subject is expected to be authenticated, such as for the purposes of logging into a website. 
```

On further thought, we could just use `authentication` as the method for allowing voting permissions. The controllers could always change the authentication method. 

My concern here is for complex cases, say if a member is actually another fora network. In this case, there may exist a slew of authentication methods (public keys for each sub-member), but authorization might be a multi-sig of the sub-members or even a passed gov proposal. Here authorization is a stronger property than authentication and if the network is to vote, you likely want the strongest verification method?

Thoughts?

File: core/activity_pub.md (hunk 13 above)
Comment (colin-axner) [reply core/activity_pub.md]: Yes, this depends on your DID implementation. [Service endpoints](https://www.w3.org/TR/did-core/#service-endpoints) are natively supported. 

Yes you can, the DID document for the community would include extra [parameters](https://www.w3.org/TR/did-core/#did-parameters) for these keys. Then if you wanted one of those values, you would enter the DID, as a DID URL
`did:cosmoshub5:vov.camp?enforcers=true`

File: core/activity_pub.md (hunk 12 above)
Comment (colin-axner) [reply core/activity_pub.md]: Let's do this. Ideally, members have a DID that defines a third verification relationship `voting authentication` or maybe `blockchain authentication`. In this situation, we use that verification relationship for verifying votes. If this verification relationship does not exist (member DIDs could use entirely different DID implementations), then we should gracefully fail on using the authorization method of the DID since I'm pretty sure that should always be there

Comment (colin-axner) [reply core/activity_pub.md]: see pushed changes

File: core/activity_pub.md (hunk 13 above)
Comment (colin-axner) [reply core/activity_pub.md]: anything I need to add to the doc?

File: core/activity_pub.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
+# Activity Pub Compliance
+
+This document specifies the extensions to Fora required for a community to be compliant with the [Activity Pub](https://www.w3.org/TR/activitypub/) specification. Without this extension, the Fora kernel will not be capable of interacting with other Activity Pub implementations.
+
+## Introduction
+
+"it's just activity pub with governance"
+
+If we examine the Fora kernel through the lens of Activity Pub we find numerous similarities. Each user is a Client and each host is a Server, as defined by the Activity Pub specification. In fact, the only differences are a layer of governance settled on top of Activity Pub, along with additional properties between clients and servers, and servers and servers. The primary property added is the right to collective identification as a single community. Different servers with different sets of clients can now collectively agree to be identified as the same community thereby sharing their state and community membership.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
If we examine the Fora kernel through the lens of Activity Pub we find numerous similarities. Each member is a Client and each host is a Server, as defined by the Activity Pub specification. In fact, the only differences are a layer of governance settled on top of Activity Pub, along with additional properties between clients and servers, and servers and servers. The primary property added is the right to collective identification as a single community. Different servers with different sets of clients can now collectively agree to be identified as the same community thereby sharing their state and community membership.
```

File: core/activity_pub.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+The usefulness of this conceptual nesting is addressing the **right to individual exit** and the **right to collective exit**. 
+
+The former can be relatively easily addressed in Activity Pub by using flexible URI's such as DID's. This would give any user power to exit a community without losing its account information (content may have to be manually ported, but followers/following lists would automatically follow the user).
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
The former can be relatively easily addressed in Activity Pub by using flexible URI's such as DID's. This would give any member the power to exit a community without losing its account information (content may have to be manually ported, but followers/following lists would automatically follow the user).
```

File: core/activity_pub.md [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+Communities will also have DIDs that identify them to their users. This DID will also be recorded on a blockchain as a government. The DID Document for this DID will be deterministically generated based on the community parameters that exist at the block height the DID resolving function is being executed.  All governance decisions will be recorded on the DID Document. The DID controller will be the community members and the verification method to update the DID Document will be the passing of a governance proposal. The DID may include a default set of "seed" hosts to allow users to conveniently find and connect to the community which will be set under the service endpoint section of a DID Document. Additional parameters will be added to the DID Document which indicate a list of admittors, enforcers, members, governance decisions and any other relevant information.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
Communities will also have DIDs that identify them to their members. This DID will also be recorded on a blockchain as a government. The DID Document for this DID will be deterministically generated based on the community parameters that exist at the block height the DID resolving function is being executed.  All governance decisions will be recorded on the DID Document. The DID controller will be the community members and the verification method to update the DID Document will be the passing of a governance proposal. The DID may include a default set of "seed" hosts to allow members to conveniently find and connect to the community which will be set under the service endpoint section of a DID Document. Additional parameters will be added to the DID Document which indicate a list of admittors, enforcers, members, governance decisions and any other relevant information. 
```

File: core/activity_pub.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+Note: There may be many valid, honest hosts that exist outside of this community-curated set.
+
+## Public/Private Key Authentication
+
+ActivityPub does not specify a particular authentication scheme, however most ActivityPub communities use OAuth2.0 or something similar which involves a traditional username/password scheme. Since all of the community state must be public and shareable between hosts, we cannot rely on username/password schemes or any scheme that gives servers privileged information on a client’s authentication info. Thus, clients must authenticate by proving ownership of a secret private key that corresponds to the public key that is publicly registered on a blockchain.
+
+## Server
+
+In addition to maintaining the full state of the community, servers will listen to events on the ActivityPub network to record and propagate new interactions. Servers must be connected to the appropriate blockchains or DID resolution services in order to track and verify the DID of users in the community.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
In addition to maintaining the full state of the community, servers will listen to events on the ActivityPub network to record and propagate new interactions. Servers must be connected to the appropriate blockchains or DID resolution services in order to track and verify the DID of members in the community.
```

File: core/activity_pub.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+## Limitations
+
+Since any content posted to a Fora community must contend with the possibility of byzantine host behaviour, the Fora protocol cannot make any guarantees regarding private posts that are only readable to a subset of users. Thus, Fora will only support Public posts and direct messages. Public posts will be sent in plaintext and signed by the sender, while direct messages to users can be end-to-end encrypted given that all users have a Public/Private key as mentioned above (though a different pair would be used for DMs). This ensures that hosts have no ability to attack the confidentiality or integrity of user-sent messages. They can only refuse to relay the messages, though the message may simply get relayed through a different path.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
Since any content posted to a Fora community must contend with the possibility of byzantine host behaviour, the Fora protocol cannot make any guarantees regarding private posts that are only readable to a subset of members. Thus, Fora will only support Public posts and direct messages. Public posts will be sent in plaintext and signed by the sender, while direct messages to members can be end-to-end encrypted given that all members have a Public/Private key as mentioned above (though a different pair would be used for DMs). This ensures that hosts have no ability to attack the confidentiality or integrity of member-sent messages. They can only refuse to relay the messages, though the message may simply get relayed through a different path.
```

File: dogma/goals.md [hunk 19]
```diff
@@ -13,12 +13,13 @@
 7. In order to efficiently and effectively enforce community rules, members of the community may appoint trusted users to specific roles that help steward the community (eg admitting new members, enforcing rules, etc). However, they may also replace these appointees at any time by voting them out and replacing with a new candidate. 
 8. Any malicious authority must not be able to use their power to suppress the ability of the community to overthrow them. For example, a moderator should not be able to ban all users that vote against them before they can vote them out.
 9. *A user who wishes to individually exit the community should be able to transfer their established identity to a different community.*
-10. *Community governance should be a liquid democracy where users can delegate their voting power to other users who can vote on their behalf*
+10. *Users in different communities should be capable of communicating with one another (given both parties consent).*
```
Comment (AdityaSripal) [inline dogma/goals.md]:
```suggestion
10. *Members in different communities should be capable of communicating with one another (given both parties consent).*
```

File: dogma/goals.md [hunk 20]
```diff
@@ -13,12 +13,13 @@
...
+11. *Community governance should be a liquid democracy where users can delegate their voting power to other users who can vote on their behalf*
```
Comment (AdityaSripal) [inline dogma/goals.md]:
```suggestion
11. *Community governance should be a liquid democracy where members can delegate their voting power to other members who can vote on their behalf*
```

File: core/core.md [hunk 21]
```diff
@@ -129,4 +129,4 @@ As noted in Non-Goal #3 in the [goals doc](../dogma/goals.md), this is not somet
 
 As mentioned above, for an honest users to successfully connect to the rest of the honest community; they must not be eclipsed by dishonest Hosts. If an honest user has no choice but to access the community through a dishonest Host, then Fora can make no guarantee that the user can be protected from banned users and banned content.
 
-Thus the Fora protocol assumes that the peer-to-peer network is sufficiently distributed and decentralized such that no user can be fully eclipsed by dishonest Hosts.
\ No newline at end of file
+Thus the Fora protocol assumes that the peer-to-peer network is sufficiently distributed and decentralized such that no user can be fully eclipsed by dishonest Hosts.
```
Comment (AdityaSripal) [inline core/core.md]:
```suggestion
Thus the Fora protocol assumes that the peer-to-peer network is sufficiently distributed and decentralized such that no member can be fully eclipsed by dishonest Hosts.
```

**Review by AdityaSripal — APPROVED** [review]
Just switching user to member where relevant

File: core/activity_pub.md [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+
+The DIDs of a client must be recorded on a verifiable data registry, typically this will be a public blockchain or ledger. All communities will rely on client DIDs to resolve the authentication method used by a particular client (public key the client signs ActivityPub messages with). Communities may choose not to rely on DIDs for authentication of governance module interaction. If the communities choose to use on-chain DID verification, the client DID controllers must be verifiable via a communication protocol such as InterBlockchain Communication. It is recommended that governance modules support using a third verification relationship known as `voting authentication` to authenticate any votes casted by a DID controller on behalf of a DID subject. If the `voting authentication` verification relationship does not exist, governance modules should default to using the authroization method of DID controllers to verify casted votes. If communities do not want to require on-chain verifiable DIDs for each member, members must register with the governance module a public key that they will use to cast votes.
```
Comment (AdityaSripal) [inline core/activity_pub.md]:
```suggestion
The DIDs of a client must be recorded on a verifiable data registry, typically this will be a public blockchain or ledger. All communities will rely on client DIDs to resolve the authentication method used by a particular client (public key the client signs ActivityPub messages with). Communities may choose not to rely on DIDs for authentication of governance module interaction. If the communities choose to use on-chain DID verification, the client DID controllers must be verifiable via a communication protocol such as InterBlockchain Communication. It is recommended that governance modules support using a third verification relationship known as `voting authentication` to authenticate any votes cast by a DID controller on behalf of a DID subject. If the `voting authentication` verification relationship does not exist, governance modules should default to using the authorization method of DID controllers to verify casted votes. If communities do not want to require on-chain verifiable DIDs for each member, members must register with the governance module a public key that they will use to cast votes.
```

File: core/activity_pub.md (hunk 12 above)
Comment (colin-axner) [reply core/activity_pub.md]: I'll put up a new pr later, but looks like we can use capabilities. It's unclear to me rn if you can create multiple capabilities with various delegates for each capability
```
The capabilityDelegation verification relationship is used to specify a mechanism that might be used by the DID subject to delegate a cryptographic capability to another party, such as delegating the authority to access a specific HTTP API to a subordinate.
An example of when this property is useful is when a DID Subject chooses to grant their capability to start a vehicle through the combined usage of a verification method and the StartCar capability to a party other than themselves. 
```
